### ISSUE TITLE
Cancelling stream before calling TarReader.moveNext()

### ISSUE THREAD

**Issue by jonasfj** [simolus3/tar#23]
Reproduction steps:
 * `dart create foo`
 * Copy/paste example code below into `lib/foo.dart`
 * Run `tar cf archive.tar lib/` to create an archive with non-ascii characters
 * Run `dart lib/foo.dart`
 * Throws `Bad state: Illegal call to TarReader.moveNext() while a previous stream was active....`

I discovered this because I was using `await utf8.decodeStream(entry.contents);` and handling `FormatException` because I wanted to ignore such files. But it seems that's not easily possible. I'm not entirely sure if this is something in `Encoding.decodeStream` or `TarReader`, I suspect `TarReader` because code for `Encoding.decodeStream` has `.listen(..., cancelOnError: true)` -- but it's hard to be sure.

**Example code**

```dart
import 'dart:convert';

import 'package:tar/tar.dart';
import 'dart:io';

// Create an archive of this file using: tar cf archive.tar lib/
// it has none ascii
// 😀

void main() async {
  final s = File('archive.tar').openRead();
  await TarReader.forEach(s, (entry) async {
    print(entry.name);
    try {
      await ascii.decodeStream(entry.contents);
    } on FormatException {
      // pass
    }
  });
}
```

Full stack trace:
```
Unhandled exception:
Bad state: Illegal call to TarReader.moveNext() while a previous stream was active.
When listening to tar contents, make sure the stream is complete or cancelled before calling TarReader.moveNext() again.
#0      TarReader._prepareToReadHeaders (package:tar/src/reader.dart:282:9)
#1      TarReader.moveNext (package:tar/src/reader.dart:116:11)
#2      TarReader.forEach (package:tar/src/reader.dart:250:27)
<asynchronous suspension>
#3      main (package:te/te.dart:10:3)
<asynchronous suspension>
```

**Comment by simolus3** [comment]
I think the problem is with `Stream.eventTransformed`: It doesn't close the transforming sink in response to cancellations. This script never prints `close on sink` for instance:

```dart
import 'dart:async';

void main() async {
  final stream = Stream<int>.eventTransformed(someInts(), _PrintOnCloseSink.new);
  final subscription = stream.listen(null);
  await subscription.cancel();
  print('cancelled');
}

Stream<int> someInts() async* {
  var i = 0;

  while (true) {
    yield i++;
    await Future<void>.delayed(const Duration(milliseconds: 100));
  }
}

class _PrintOnCloseSink extends EventSink<int> {
  final EventSink<int> _inner;

  _PrintOnCloseSink(this._inner);

  @override
  void add(int event) {
    _inner.add(event);
  }

  @override
  void addError(Object error, [StackTrace? stackTrace]) {
    _inner.addError(error, stackTrace);
  }

  @override
  void close() {
    print('close on sink');
    _inner.close();
  }
}
```

That said, it looks like I was aware of this and just didn't consider the case of cancelling the stream:

https://github.com/simolus3/tar/blob/901ae404e0a225d9b08e5253415ca092f5c08706/lib/src/reader.dart#L464-L465

So it seems like I need to avoid `eventTransformed` here. In response to a cancellation request, we could either

1. Immediately consume the rest of the entries stream before `StreamSubscription.cancel` completes.
2. Complete `cancel` immediately and consume the rest of the current entry once `hasNext` is called again.

I think the second approach is cleaner here. It matches the wisdom shared in https://github.com/simolus3/tar/issues/6#issuecomment-756726130. The first approach also leaves potential for a `cancel` call to throw an exception if the source stream emits an error just after an entry has been cancelled.

**Comment by simolus3** [comment]
Fixed in `package:tar` version 0.5.6! Do you want me to roll that into pub as well? It shouldn't affect anything pub is doing, right?

**Comment by jonasfj** [comment]
Rolling into pub is certainly nice! :D

I don't think it causes any issues, but there is no reason to let it diverge.
